json-path/src/main/java/com/jayway/jsonpath/Predicate.java

```diff
@@ -23,6 +24,8 @@
 
     boolean apply(PredicateContext ctx);
 
+    boolean check(TokenStack stack, int idx);
```

Comment:
So it seems there are a couple of options:
1) no streaming -- but there are JsonPath users who do want streaming2) split the predicate interface into two interfaces3) add a small class that takes a Predicate and calls check() ie this would be a "proxy"/"facade" that is functional and the Predicate interface remains as the logical interface to uniary and binary operators4) give up on lambdas
Not sure I understand why you care about lambdas so much in a non-streaming context.  Lambdas are just syntactic sugar unless you are doing something that needs streaming ie the DOM is too big for memory.  In the non-streaming case I doubt the difference in performance between a lambda and an anonymous inner class would matter.  I guess it all about Scala probably though huh...
Its probably bad design to tie the Predicate interface to both the way to call predicates on a "parsing context" and as the way to extend the logical operators JsonPath supports (AND, NOT, OR, etc).
Either way, I would feel more comfortable if someone else made this decision for several reasons including the fact that working on this project is a low priority task for me.  My company doesn't make parsers, we make streaming DB engines.  We just needed a streaming XPath like thing for JSON and this project seems to be the closest to that today.  If something else was available, we would use that instead and this pull request wouldn't exist.
Hunter
PS On a side note, its extra asks like this that make me less likely to donate code to open projects in the future.  I'm trying to do the right thing and give back to the project and all I get for that decision is extra work to do.  There was no way I could know you wanted to functional interfaces in JsonPath when I was writing this bit of code.  Adding comments to that effect in JsonPath would be very helpful.  I apologize if you did that and I overlooked it.PSS I've been working with open source projects for 20 years and donated 100,000s LOC to various projects so I feel like I've already paid that debt so to speak for all the libraries I've included in my commercial work over the years.
 


     On Tuesday, July 28, 2015 12:58 AM, Claus Ibsen <notifications@github.com> wrote:
   

 In json-path/src/main/java/com/jayway/jsonpath/Predicate.java:> @@ -23,6 +24,8 @@
>  
>      boolean apply(PredicateContext ctx);
>  
> +    boolean check(TokenStack stack, int idx);
I think he means that the interface should only have 1 method, which is what allows the use with lambdas in the Java 8 onwards. eg so you cannot add a new check method.—
Reply to this email directly or view it on GitHub.